### PR TITLE
rulesets for title, image, price; 

### PR DESCRIPTION
**Pull request by swathiiyer2**
-I've run the cpu profiler 
-Refactored code to do file reading, map creating at the beginning
-Accuracy as of now: title 100%, image 90%, price 92%
-Next step: new feature? improve existing?

### CODE REVIEW

File: products.js
Comment (erikrose) [inline products.js]: Fortunately you're multiplying 2 negatives together, which cancels out the error. :-)

File: product_classification_test_data/allposters/source.html
Comment (erikrose) [inline product_classification_test_data/allposters/source.html]: These source (and nodes) files change more than I would expect. For instance, they just changed 2 commits ago, in 69edc4a5c557a6fea3b142321d4a9b9120dcb7fe. Are we still correcting something?

File: products.js
Comment (erikrose) [reply products.js]: Should probably fix that, just the same :-)

File: products.js
Comment (erikrose) [inline products.js:15]: I'm generally confused by this commit. The commit message says this commit is all about splitting the corpus, but there appear to be a lot of changes here that have nothing to do with that. In the future, could you could break up your commits or at least outline all the changes in the commit message?

File: products.js
Comment (erikrose) [inline products.js]: Ah, are these the cause of the source.html changes?

File: products.js
Comment (erikrose) [inline products.js]: `coefftitlewords` doesn't follow the casing conventions.

Comment (erikrose) [reply products.js]: Nor does titlewordsBase.

File: products.js
Comment (erikrose) [inline products.js]: The usual idiom is to leave a space after `for`, `if`, etc.

File: products.js
Comment (erikrose) [inline products.js]: Hmm! So you're propping up the coeffs with multiplication by a constant. Why not factor those into the coeffs themselves? Then I wouldn't be mislead about their relative weights when reading them. (You can and should factor those up even before running the optimizer. The optimizer should be able to reach any number.)

File: products.js
Comment (erikrose) [inline products.js]: It's Super Bad™ to have 2 symbols floating around that differ only in case. :-o In general, anything that makes the reader flip to a different line to figure out what you meant is nonoptimal.

File: products.js
Comment (erikrose) [inline products.js]: Another really good contribution (since you're looking for something to do) would be to figure out what primitives are hiding behind these couple of similar attribute loops you have. What could Fathom do to relieve you and Future Yous of writing these arduous loops?

File: products.js
Comment (erikrose) [inline products.js]: Did you actually hit SVGs in real life? Curious.

File: products.js
Comment (erikrose) [inline products.js]: I wonder if css.top can go negative.

File: products.js
Comment (erikrose) [inline products.js]: Pretty dang cool. Excited to be using spatiality. :-)

File: products.js
Comment (erikrose) [inline products.js]: Shouldn't this be called notDataURLs for consistency with notSVGs?

File: products.js
Comment (erikrose) [inline products.js]: I see this optimized out to 0.5. I wonder if it really does score better than if you make it 0.

File: products.js
Comment (erikrose) [inline products.js]: This and the above includes() call would be a good thing to factor out into a withoutQueryString() routine. Then you could avoid linearly scanning through the whole string twice for the question mark while still staying consise at the callsite. As a bonus, the name of the new routine would make it even clearer what you're doing.

File: products.js
Comment (erikrose) [inline products.js]: This comment is now wrong.

File: products.js
Comment (erikrose) [inline products.js]: A lot of these numbers are probably equal to their shorter representations, like 1.9 and 3.4. Floating point has only so much resolution, so you tend to get artifacts like this. (After all, your tuning routine adds or subtracts only tenths at a time, so how could it be otherwise?) We can thus shorten the spelling here for the poor humans who get to read it. :-)

File: products.js
Comment (erikrose) [inline products.js]: You know, you wrote this ruleset in a way that I never would have dared. And it totally worked! You used expensive constructs like .textContent (which has to iteratate over a whole subtree and collect the text) with abandon, and you got a very high score. My congratulations!

But that does bring to mind another thing you can do (since you were looking): bench your rulesets and see if they take more than a couple milliseconds. If they do and there's an evident way to speed them up, do so. That's another point of reluctance I've heard in some quarters about smarter browser behavior.

File: products.js
Comment (erikrose) [inline products.js]: On the subject of making the corpus harder, add some non-US-currency pages to your list of things to go find.

File: products.js
Comment (erikrose) [inline products.js]: All these callbacks will benefit from the predicate functionality I mean to attach to the LHS, e.g. `dom('whatever').when(fnode => fnode.element.tagName === 'SPAN')`. Thanks for surfacing the pattern!

File: products.js
Comment (erikrose) [inline products.js]: I think you're missing a `.length` on the second condition.

File: products.js
Comment (erikrose) [inline products.js]: Another spot that will face problems on non-English pages. That's another direction you could attack.

File: products.js
Comment (erikrose) [inline products.js]: It sounds like an in-ruleset primitive for "if this DOM element we've already selected matches this CSS selector" would be useful. Do you agree?

Btw, let's factor out the repeated class/id word lists.

File: products.js
Comment (erikrose) [inline products.js]: It looks like you're blanking out any prices you find. Doesn't that defeat the point, or am I missing something? Probably the latter; please fill me in.

File: products.js
Comment (erikrose) [inline products.js]: If there's reason to keep this price-blanking stuff, we at least shouldn't write it 3 times. DRY!

**Review by erikrose — CHANGES_REQUESTED** [review]
Okay, I made a big pile of comments, because it was a big pile of code. Let's do smaller piles from now on. Shorter commits are easier to think about, easier to review and talk about, and easier to rework. That's especially true if each commit passes the tests as they exist at that time, though that's not possible or desirable every single time.

Things I'd like you to do:
* Add more commits (one per comment would be a good start, if you can't think of another sensical way to organize them) that address my comments. (Of course, when I go astray, argue. :-))
* Collect the Possible Next Steps from my comments and from my email yesterday, and mail them to me. I speak of things like "You could work on non-English pages. Or non-dollar currencies." We'll talk through them when you're done and pick what to do.
* Before you embark on whatever we pick, I think the most important thing is actually a cleaning pass. Apply my suggestions from this PR, shorten code, and just generally compact and clarify. Make sure comments are correct and up to date. Make everything well-organized and easy to read. You'll need something understandable to put on a slide (or 2 or 3), and people will have to "get it" at a glance. So spend some time polishing. In addition to having a nice impact on your presentation, this will make it easier and more pleasant to work in here, and your code will live on, rather than being relegated to the pile of "things an intern wrote and we abandoned". I think this really might go somewhere if its grizzled appearance doesn't turn future maintainers off.

Good work, Swathi. I honestly didn't expect you to get it scoring so well, especially not so early. There are rough spots in here, but it really does work! Let's clean it up, collect ourselves, and figure out in which direction to strike next. Good night!

File: products.js
Comment (swathiiyer2) [reply products.js]: Yes!

File: products.js
Comment (swathiiyer2) [reply products.js]: Will do! I also tried out another set of coefficients for prices that I got from the optimizer that give the same accuracy on the training data, and is giving 4% better accuracy, I could use this instead? (will be at 94%)

Comment (erikrose) [reply products.js]: Sure, why not?

File: products.js
Comment (swathiiyer2) [reply products.js]: I don't think so. getBoundingClientRect() gives distance from the top of window, which starts at 0.

File: products.js
Comment (swathiiyer2) [reply products.js]: Agree, that would've very much helped. Will fix!

File: products.js
Comment (swathiiyer2) [reply products.js]: I'm actually just removing the dollar sign, stripping whitespace, and removing any trailing zeroes to standardize the format/avoid false negatives (I am getting the actual prices). I can try to condense this!

Comment (erikrose) [reply products.js]: Ah, I didn't see the `^` in my haste. But `'59.7700'.replace(/(\.[0-9]*?)0+$/, '');` is `59`. You didn't mean to strip off the 77¢, did you?

Comment (swathiiyer2) [reply products.js]: Oops I didn't mean to do that. Do you know what the best way would be to remove trailing zeroes?

Comment (swathiiyer2) [reply products.js]: never mind, got it!

**Comment by erikrose** [comment]
Looks good. I can't think of an alternative off the top of my head, but keep in mind that in general the non-greedy quantifiers (`*?`, `+?`) can get really, really, *really* slow, so reach for negated character classes (`[^a-z]`, etc.) when possible.

**Comment by erikrose** [comment]
Great small commits. This is exactly what I was hoping for!

File: products.js
Comment (erikrose) [inline products.js]: Do you mean to comment these in and out on every commit? There's a lot of thrash. I suspect you are not reading your diffs before committing. :-o ;-)

File: products.js
Comment (erikrose) [inline products.js]: In the case of a substantial change like this, I'd want to see a rationale in the commit message. This helps the reviewer and anybody who runs `git blame` someday.

File: products.js
Comment (erikrose) [inline products.js]: You use this pattern in a few places: includes() and then a substr(). See if there's already a string method for this. If not, here's a great opportunity to make one. It'll make your code clearer and probably boost performance, since you won't have to linear-scan through the string twice.

File: products.js
Comment (erikrose) [inline products.js]: A good factoring up :-)

File: products.js
Comment (erikrose) [inline products.js]: You could even factor formatPrice out of the ternary expression.

File: products.js
Comment (erikrose) [inline products.js]: Is this change on purpose? (It wasn't mentioned in the commit message, and I can't tell offhand what it does, so I'm not sure.)

File: products.js
Comment (swathiiyer2) [reply products.js]: oh I didn't mean to, will remember to check now!

File: products.js
Comment (swathiiyer2) [reply products.js]: Okay! I just removed it to match the effect of the forgotten .length I had before. (adding the .length actually worsened the accuracy)

File: products.js
Comment (swathiiyer2) [reply products.js]: Oh, I just combined two of the regexes!

File: products.js
Comment (swathiiyer2) [reply products.js]: Ohh yes I shouldn't need the includes at all, because indexOf returns -1.

**Comment by erikrose** [comment]
Ah, interesting!

File: products.js
Comment (erikrose) [reply products.js]: Ah, interesting!

File: products.js
Comment (erikrose) [inline products.js]: This just returns "" if there's no "?" in `url`.

File: products.js
Comment (erikrose) [inline products.js]: I'm not sure how this makes sense either; there certainly won't be any "$" in `priceString` once you've run `replace('$', '')`.

File: products.js
Comment (swathiiyer2) [reply products.js]: Ooh, guess I'd better change this back then.

File: products.js
Comment (swathiiyer2) [reply products.js]: I actually ran into a few instances where there would be multiple of the same price in the text, but they're still both correct

Comment (swathiiyer2) [inline products.js]: How would I account for the above case, as well a case where it gives me something like 199.99$ (the price is correct, but there's a dollar sign at the end) with just one pass through the string?

Comment (erikrose) [reply products.js]: I'm sorry; I was imposing the Python replace semantics onto JS. In Python, it's always a global replace.

File: products.js
Comment (erikrose) [inline products.js]: Didn't you mean indexOf?

Comment (swathiiyer2) [reply products.js]: Oh man, keep making mistakes on this sorry!!

File: products.js
Comment (erikrose) [inline products.js:29]: I don't see how trailing zeros are stripped. Is this comment still accurate?

Comment (swathiiyer2) [reply products.js:29]: Trailing zeros are stripped when I call parseFloat()

File: products.js
Comment (erikrose) [reply products.js]: Read your diffs and run the tests before committing. Always. :-) (If you don't have tests, test it manually. Then the pain will drive you to write tests.)

File: products.js
Comment (erikrose) [reply products.js:29]: So they are! Thanks for spelling it out for me.

File: products.js
Comment (swathiiyer2) [reply products.js]: Ohh I see no worries!

**Review by erikrose — APPROVED** [review]

**Comment by erikrose** [comment]
Approved! Would have been sooner if GH had bothered to send me a mail. :-(

**Pull request merged**
